diff --git a/auto/lib/openssl/make b/auto/lib/openssl/make
index 3af2ae55..126a2387 100644
--- a/auto/lib/openssl/make
+++ b/auto/lib/openssl/make
@@ -49,13 +49,11 @@ END
         cat << END                                            >> $NGX_MAKEFILE
 
 $OPENSSL/.openssl/include/openssl/ssl.h:	$NGX_MAKEFILE
-	mkdir -p $OPENSSL/build $OPENSSL/.openssl/lib $OPENSSL/.openssl/include/openssl \\
-	&& cd $OPENSSL/build \\
-	&& cmake -DCMAKE_C_FLAGS="$OPENSSL_OPT" -DCMAKE_CXX_FLAGS="$OPENSSL_OPT" .. \\
-	&& \$(MAKE) VERBOSE=1 \\
-	&& cd .. \\
-	&& cp -r include/openssl/*.h .openssl/include/openssl \\
-	&& cp build/ssl/libssl.a build/crypto/libcrypto.a .openssl/lib
+	cd $OPENSSL \\
+	&& if [ -f Makefile ]; then \$(MAKE) clean; fi \\
+	&& ./config --prefix=$ngx_prefix no-shared no-threads $OPENSSL_OPT \\
+	&& \$(MAKE) \\
+	&& \$(MAKE) install_sw LIBDIR=lib
 
 END
 
