diff --git a/auto/lib/openssl/conf b/auto/lib/openssl/conf
index f0dd1e1..9d9afb0 100644
--- a/auto/lib/openssl/conf
+++ b/auto/lib/openssl/conf
@@ -144,39 +144,11 @@ fi
 
 if [ $USE_OPENSSL_QUIC = YES ]; then
 
-    ngx_feature="OpenSSL QUIC support"
-    ngx_feature_name="NGX_OPENSSL_QUIC"
-    ngx_feature_run=no
-    ngx_feature_incs="#include <openssl/ssl.h>"
-    ngx_feature_path=
-    ngx_feature_libs="-lssl -lcrypto $NGX_LIBDL $NGX_LIBPTHREAD"
-    ngx_feature_test="SSL_set_quic_method(NULL, NULL)"
-    . auto/feature
-
-    if [ $ngx_found = no ]; then
-
-cat << END
-
-$0: error: certain modules require OpenSSL QUIC support.
-You can either do not enable the modules, or install the OpenSSL library
-into the system, or build the OpenSSL library statically from the source
-with nginx by using --with-openssl=<path> option.
-
-END
-        exit 1
-    fi
-
+    have=NGX_OPENSSL_QUIC . auto/have
     have=NGX_QUIC . auto/have
 fi
 
 
 if [ $USE_OPENSSL_QUIC = YES ]; then
-    ngx_feature="OpenSSL QUIC 0-RTT context"
-    ngx_feature_name="NGX_OPENSSL_QUIC_ZRTT_CTX"
-    ngx_feature_run=no
-    ngx_feature_incs="#include <openssl/ssl.h>"
-    ngx_feature_path=
-    ngx_feature_libs="-lssl -lcrypto $NGX_LIBDL $NGX_LIBPTHREAD"
-    ngx_feature_test="SSL_set_quic_early_data_context(NULL, NULL, 0)"
-    . auto/feature
+    have=NGX_OPENSSL_QUIC_ZRTT_CTX . auto/have
 fi
